<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        <meta name="author" content="shenwei356">
        
        <link rel="shortcut icon" href="../img/favicon.ico">

	<title>Tutorial - csvtk - CSV/TSV Toolkit</title>

        <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
        <link href="../css/font-awesome-4.5.0.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link rel="stylesheet" href="../css/highlight.css">

        <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
        <!--[if lt IE 9]>
            <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
            <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
        <![endif]-->

        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-28948626-3', 'blog.shenwei.me/csvtk');
            ga('send', 'pageview');
        </script>
        
    </head>

    <body>

        <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->
            <a class="navbar-brand" href="..">csvtk - CSV/TSV Toolkit</a>
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
            
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../download/">Download</a>
                    </li>
                
                
                
                    <li >
                        <a href="../usage/">Usage</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Tutorial</a>
                    </li>
                
                
                </ul>
            

            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                        <i class="fa fa-search"></i> Search
                    </a>
                </li>
                
                    <li >
                        <a rel="next" href="../usage/">
                            <i class="fa fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li class="disabled">
                        <a rel="prev" >
                            Next <i class="fa fa-arrow-right"></i>
                        </a>
                    </li>
                
                
                    <li>
                        <a href="https://github.com/shenwei356/csvtk">
                            
                                <i class="fa fa-github"></i>
                            
                            GitHub
                        </a>
                    </li>
                
            </ul>
        </div>
    </div>
</div>

        <div class="container">
            
                <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
    
        <li class="main active"><a href="#tutorial">Tutorial</a></li>
        
            <li><a href="#forewords">Forewords</a></li>
        
            <li><a href="#analyze-otu-table">Analyze OTU table</a></li>
        
    
    </ul>
</div></div>
                <div class="col-md-9" role="main">

<h1 id="tutorial">Tutorial</h1>
<h2 id="forewords">Forewords</h2>
<p>Yes, you could just use spreadsheet softwares like MS excel to
do most of the job.</p>
<p>Howerver it's all by clicking and typing, which is <strong>not
automatically and time-consuming to repeate</strong>, especially when we want to
apply similar operations with different datasets or purposes.</p>
<p><code>csvtk</code> is <strong>convenient for rapid investigation
and also easy to integrated into analysis pipelines</strong>.
 It could save you much time of writting scripts.</p>
<p>Hope it be helpful for you.</p>
<h2 id="analyze-otu-table">Analyze OTU table</h2>
<h3 id="data">Data</h3>
<p>Here is mock a OTU table from 16S sequencing result.
Columns are sample IDs in format of "GROUP.ID"</p>
<pre><code>    $ cat otu_table.csv
    Taxonomy,A.1,A.2,A.3,B.1,B.2,B.3,C.1,C.2,C.3
    Proteobacteria,.13,.29,.13,.16,.13,.22,.30,.23,.21
    Firmicutes,.42,.06,.49,.41,.55,.41,.32,.38,.66
    Bacteroidetes,.19,.62,.12,.33,.16,.29,.34,.35,.09
    Deferribacteres,.17,.00,.24,.01,.01,.01,.01,.01,.02
    Tenericutes,.00,.00,.00,.01,.03,.02,.00,.00,.00
</code></pre>
<h3 id="steps">Steps</h3>
<ol>
<li>
<p>Counting</p>
<pre><code>$ csvtk stat otu_table.csv
file            num_cols   num_rows
otu_table.csv         10          5
</code></pre>
</li>
<li>
<p>Convert to tab-delimited table</p>
<pre><code>$ csvtk csv2tab  otu_table.csv
Taxonomy        A.1     A.2     A.3     B.1     B.2     B.3     C.1     C.2     C.3
Proteobacteria  .13     .29     .13     .16     .13     .22     .30     .23     .21
Firmicutes      .42     .06     .49     .41     .55     .41     .32     .38     .66
Bacteroidetes   .19     .62     .12     .33     .16     .29     .34     .35     .09
Deferribacteres .17     .00     .24     .01     .01     .01     .01     .01     .02
Tenericutes     .00     .00     .00     .01     .03     .02     .00     .00     .00
</code></pre>
</li>
<li>
<p>Column names</p>
<pre><code>$ csvtk cut -n otu_table.csv
#field  colname
1       Taxonomy
2       A.1
3       A.2
4       A.3
5       B.1
6       B.2
7       B.3
8       C.1
9       C.2
10      C.3
</code></pre>
</li>
<li>
<p>Extract data of group A and B and save to file <code>-o otu_table.gAB.csv</code></p>
<pre><code>$ csvtk cut -F -f "A.*,B.*,Taxonomy" otu_table.csv -o otu_table.gAB.csv
$ cat otu_table.gAB.csv
Taxonomy,A.1,A.2,A.3,B.1,B.2,B.3
Proteobacteria,.13,.29,.13,.16,.13,.22
Firmicutes,.42,.06,.49,.41,.55,.41
Bacteroidetes,.19,.62,.12,.33,.16,.29
Deferribacteres,.17,.00,.24,.01,.01,.01
Tenericutes,.00,.00,.00,.01,.03,.02
</code></pre>
</li>
<li>
<p>Transpose</p>
<pre><code>$ csvtk transpose otu_table.gAB.csv -o otu_table.gAB.t.csv
$ csvtk csv2tab  otu_table.gAB.t.csv         
Taxonomy        Proteobacteria  Firmicutes      Bacteroidetes   Deferribacteres Tenericutes
A.1     .13     .42     .19     .17     .00
A.2     .29     .06     .62     .00     .00
A.3     .13     .49     .12     .24     .00
B.1     .16     .41     .33     .01     .01
B.2     .13     .55     .16     .01     .03
B.3     .22     .41     .29     .01     .02
</code></pre>
</li>
<li>
<p>Rename first column</p>
<pre><code>$ csvtk rename -f 1 -n "sample" otu_table.gAB.t.csv -o otu_table.gAB.t.r.csv
$ csvtk csv2tab  otu_table.gAB.t.r.csv
sample  Proteobacteria  Firmicutes      Bacteroidetes   Deferribacteres Tenericutes
A.1     .13     .42     .19     .17     .00
A.2     .29     .06     .62     .00     .00
A.3     .13     .49     .12     .24     .00
B.1     .16     .41     .33     .01     .01
B.2     .13     .55     .16     .01     .03
B.3     .22     .41     .29     .01     .02
</code></pre>
</li>
<li>
<p>Add group column</p>
<pre><code>$ csvtk mutate -p "(.+?)\." -n group otu_table.gAB.t.r.csv -o otu_table2.csv
$ csvtk csv2tab otu_table2.csv
sample  Proteobacteria  Firmicutes      Bacteroidetes   Deferribacteres Tenericutes     group
A.1     .13     .42     .19     .17     .00     A
A.2     .29     .06     .62     .00     .00     A
A.3     .13     .49     .12     .24     .00     A
B.1     .16     .41     .33     .01     .01     B
B.2     .13     .55     .16     .01     .03     B
B.3     .22     .41     .29     .01     .02     B
</code></pre>
</li>
<li>
<p>Rename groups:</p>
<pre><code>$ csvtk replace -f group -p "A" -r "Ctrl" otu_table2.csv | csvtk replace -f group -p "B" -r "Treatment" &gt; otu_table3.csv
$ csvtk csv2tab otu_table3.csv sample  Proteobacteria  Firmicutes      Bacteroidetes   Deferribacteres Tenericutes     group
A.1     .13     .42     .19     .17     .00     Ctrl
A.2     .29     .06     .62     .00     .00     Ctrl
A.3     .13     .49     .12     .24     .00     Ctrl
B.1     .16     .41     .33     .01     .01     Treatment
B.2     .13     .55     .16     .01     .03     Treatment
B.3     .22     .41     .29     .01     .02     Treatment
</code></pre>
</li>
<li>
<p>Statistics of number data</p>
<pre><code>$ csvtk stat2 -f 2-6 otu_table3.csv
field             num    sum    min    max   mean   stdev
Proteobacteria      6   1.06   0.13   0.29   0.18    0.07
Firmicutes          6   2.34   0.06   0.55   0.39    0.17
Bacteroidetes       6   1.71   0.12   0.62   0.28    0.18
Deferribacteres     6   0.44      0   0.24   0.07     0.1
Tenericutes         6   0.06      0   0.03   0.01    0.01
</code></pre>
<p>Please don't be supprised at the <code>sum</code>, it's a mock otu table...</p>
</li>
<li>
<p>Sort by abundance of <em>Proteobacteria</em> in descending order.</p>
<pre><code>$ csvtk sort -k Proteobacteria:nr otu_table3.csv -T
sample  Proteobacteria  Firmicutes      Bacteroidetes   Deferribacteres Tenericutes     group
A.2     .29     .06     .62     .00     .00     Ctrl
B.3     .22     .41     .29     .01     .02     Treatment
B.1     .16     .41     .33     .01     .01     Treatment
B.2     .13     .55     .16     .01     .03     Treatment
A.3     .13     .49     .12     .24     .00     Ctrl
A.1     .13     .42     .19     .17     .00     Ctrl
</code></pre>
</li>
<li>
<p>Sort by abundance of <em>Proteobacteria</em> in descending order and <em>Firmicutes</em> in ascending order</p>
<pre><code>$ csvtk sort -k Proteobacteria:nr -k Firmicutes:n otu_table3.csv -T
sample  Proteobacteria  Firmicutes      Bacteroidetes   Deferribacteres Tenericutes     group
A.2     .29     .06     .62     .00     .00     Ctrl
B.3     .22     .41     .29     .01     .02     Treatment
B.1     .16     .41     .33     .01     .01     Treatment
A.1     .13     .42     .19     .17     .00     Ctrl
A.3     .13     .49     .12     .24     .00     Ctrl
B.2     .13     .55     .16     .01     .03     Treatment
</code></pre>
</li>
</ol>
<div id="disqus_thread"></div>

<script>
/**
* RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
* LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables
*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL; // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');

s.src = '//csvtk.disqus.com/embed.js';

s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>

<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript" rel="nofollow">comments powered by Disqus.</a></noscript></div>
            
        </div>

        <footer class="col-md-12">
            <hr>
            
            <p>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>

        <script src="../js/jquery-1.10.2.min.js"></script>
        <script src="../js/bootstrap-3.0.3.min.js"></script>
        <script src="../js/highlight.pack.js"></script>
        <script>var base_url = '..';</script>
        <script data-main="../mkdocs/js/search.js" src="../mkdocs/js/require.js"></script>
        <script src="../js/base.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="Search Modal" aria-hidden="true">
            <div class="modal-dialog">
                <div class="modal-content">
                    <div class="modal-header">
                        <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
                        <h4 class="modal-title" id="exampleModalLabel">Search</h4>
                    </div>
                    <div class="modal-body">
                        <p>
                            From here you can search these documents. Enter
                            your search terms below.
                        </p>
                        <form role="form">
                            <div class="form-group">
                                <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query">
                            </div>
                        </form>
                        <div id="mkdocs-search-results"></div>
                    </div>
                    <div class="modal-footer">
                    </div>
                </div>
            </div>
        </div>

    </body>
</html>
